### ISSUE TITLE
Exposing headers of the CONNECT response

### ISSUE THREAD

**Issue by yutakahirano** [w3c/webtransport#264]
It may be useful to expose the headers of the CONNECT response.

This is related to #263. While we may want to filter some headers (`set-cookie`, for example?), this feature is less dangerous than #263.

**Comment by jan-ivar** [comment]
The priority of this seems tied to https://github.com/w3c/webtransport/issues/263.

**Comment by jan-ivar** [comment]
Meeting:
- We should have a consistent position here with #263 (not many use cases of this without the other)

**Comment by wilaw** [comment]
Changing milestone as this is tied to #263 which is assigned to Candidate Recommendation.

**Comment by wilaw** [comment]
Comments from IETF #122 meeting

#264: Expose CONNECT response headers: Should allow adding headers for CONNECT request - some support online

> EK: “Why not?”, people might send headers after the connection is established, so we should allow sending headers at the CONNECT request.
> VV: Previous opinion: send in payload, now changed to headers are received before application data, so this is good. Allows pren
> MT: if we allow headers, we should also allow cookies, which open the CORS question, we currently engage with Fetch at a lower level and this would entail a higher-level engagement - this is not really a question that the IETF should be answering; the CORS question is very relevant here.
> VV: <missed what Victor said, check audio>
> chairs: question is really for the W3C to resolve

**Comment by martinthomson** [comment]
The more I think about this, the more that I think we need to just do CORS or something like it.  We could have different defaults (no credentials, therefore no preflight), but the ability to set headers is an oft-cited reason NOT to use websockets.

**Comment by martinthomson** [comment]
Yes, the point of CORS preflight is to check whether it acceptable to make the request, which includes verifying that the target resource understands the protocol.

The problem is that the request might not be received by a resource that understands webtransport, or even CORS.  An attacker might point the request at some unwitting resource.  What might be done with a CONNECT that includes cookies and arbitrary headers is probably limited in terms of concrete attacks, but the point of CORS is to shield those resources, not the ones that already understand what is going on.

**Comment by vasilvv** [comment]
If we are able to send a WebTransport CONNECT request to the server, then the server has already set appropriate HTTP/3 SETTINGS, meaning it knows what's going on.

**Comment by martinthomson** [comment]
Fair, but is that true for all resources on the server, or just the server?  Do we expect that a server advertising the setting has checked that all the resources are OK?  That might be reasonable.

**Comment by vasilvv** [comment]
It should be, unless the server somehow converts WebTransport CONNECT requests into something that is non-CONNECT-shaped.

**Comment by LPardue** [comment]
In my experience withbWebSockets, an intermediary is probably going to be configured to allow a WebTransport feature on an SNI basis and pass the request through to the origin for it to decide if the resource can handle WebTransport or not for real. 

Such an intermediary that follows the RFC will itself first validate the target origin server is capable of speaking WebTransport before issuing the CONNECT. If they chose to skip that step, then they shot themselves and put their customers at risk. Since we are not offering any WebTransport over HTTP/1.1, that limits such a server from doing a spec compliant potentially insecure transform. 

So I tend to agree with Victor that this sounds reasonable.

Intermediary transform strategies are common deployment pattern. We might consider documenting a security consideration somewhere about what might be at risk.

**Comment by martinthomson** [comment]
Consider me convinced that this doable from an application-security perspective.  The trick then is in ensuring that our use of Fetch is at the appropriate level to unlock all of the cookies and fetch markings.  Right now, we're hooking in a too low a level to get many of the Fetch features, which this would then need.

**Comment by wilaw** [comment]
> The trick then is in ensuring that our use of Fetch is at the appropriate level to unlock all of the cookies and fetch markings. Right now, we're hooking in a too low a level to get many of the Fetch features, which this would then need.

See issue #625. @jan-ivar agreed yesterday to generate a PR to tie in to FETCH at a higher level. Header support will be inherited, but several features (redirects, credentials?, CORs?) will need to be turned off. I encourage this group to review that PR once its ready.
